File: src/interpreter/token.rs
use std::{format};

#[derive(Debug)]
pub enum TokenType {
	INT,
	FLOAT,

	PLUS,
	MINUS,
	MUL,
	DIV,
	LPARENT,
	RPARENT
}

impl TokenType {
	pub fn from_char(c: char) -> Option<Self> {
		match c {
			'+' => Some(TokenType::PLUS),
			'-' => Some(TokenType::MINUS),
			'*' => Some(TokenType::MUL),
			'/' => Some(TokenType::DIV),
			'(' => Some(TokenType::LPARENT),
			')' => Some(TokenType::RPARENT),
			_ => None
		}
	}
}



#[derive(Debug)]
pub enum TokenValue {
	// UNDEFINED,
	IntegerValue(i32),
	FloatValue(f64),
	StringValue(String)
}

impl TokenValue {
	pub fn to_string(&self) -> String {
		match self {
			// TokenValue::UNDEFINED => "UNDEFINED".to_string(),
			TokenValue::IntegerValue(value) => value.to_string(),
			TokenValue::FloatValue(value) => value.to_string(),
			TokenValue::StringValue(value) => value.clone(),
		}
	}
}



#[derive(Debug)]
pub struct Token {
	ttype: TokenType,
	value: TokenValue
}


impl Token {
	pub fn new(ttype: TokenType, value: TokenValue) -> Self {
		Token { ttype, value }
	}

	pub fn as_string(&self) -> String {
		format!("{:?}:{}", self.ttype, self.value.to_string())
	}
}

pub fn token_vector_as_string(tokens: &Vec<Token>) -> Vec<String> {
	let mut token_str: Vec<String> = Vec::new();

	for token in tokens {
		token_str.push(token.as_string());
	}

	token_str
}

File: src/interpreter/token.cpp
#include "../headers/token.hpp"
#include <optional>

// Token::Token(TokenType type, std::string value) : type(type), value(value) {}
Token::Token(TokenType type, std::string value, Position pos) : type(type), value(value), pos(pos) {}


std::string Token::as_string() {
	return get_token_string(type) + ":" + value;
}

TokenType Token::from_char(char c) {
	switch (c) {
		case '+': return TokenType::PLUS;
		case '-': return TokenType::MINUS;
		case '*': return TokenType::MUL;
		case '/': return TokenType::DIV;
		case '(': return TokenType::LPARENT;
		case ')': return TokenType::RPARENT;
		default: return TokenType::UNKNOWN;
	}
}



TokenType Token::get_type() const {
	return type;
}

std::string Token::get_value() const {
	return value;
}

Position Token::get_pos() {
	return pos;
}


/* Private */
std::string Token::get_token_string(TokenType type) {
	switch (type) {
		// case TokenType::UNDEFINED: return "UNDEFINED";
		case TokenType::INT: return "INT";
		case TokenType::FLOAT: return "FLOAT";
		case TokenType::PLUS: return "PLUS";
		case TokenType::MINUS: return "MINUS";
		case TokenType::MUL: return "MUL";
		case TokenType::DIV: return "DIV";
		case TokenType::LPARENT: return "LPARENT";
		case TokenType::RPARENT: return "RPARENT";
		case TokenType::TEOF: return "EOF";
		default: return "Unkown";
	}
}